### PR TITLE
[Merged by Bors] - feat(Algebra/CharP/Two): more lemmas on char two

### PR DESCRIPTION
**Pull request by vihdzp**
Co-authored-by: Ruben Van de Velde <ruben.vandevelde+github@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/CharP/Two.lean [hunk 1]
```diff
@@ -20,15 +20,34 @@ variable {R ι : Type*}
...
 
+section AddMonoidWithOne
+
+variable [AddMonoidWithOne R]
+
+theorem two_eq_zero [CharP R 2] : (2 : R) = 0 := by
+  rw [← Nat.cast_two, CharP.cast_eq_zero]
+
+/-- The only hypotheses required to build a `CharP R 2` instance are `1 ≠ 0` and `2 = 0`. -/
+protected theorem mk (h₁ : (1 : R) ≠ 0) (h₂ : (2 : R) = 0) : CharP R 2 := by
```
Comment (Command-Master) [inline Mathlib/Algebra/CharP/Two.lean:31]: I'm not sure about this name, `mk` is usually the constructor of a structure.

File: Mathlib/Algebra/CharP/Two.lean [hunk 2]
```diff
@@ -37,16 +56,27 @@ variable [Ring R] [CharP R 2]
...
+
+theorem add_eq_iff_eq_add {a b c : R} : a + b = c ↔ a = c + b := by
+  rw [← sub_eq_iff_eq_add, sub_eq_add]
+
+theorem eq_add_iff_add_eq {a b c : R} : a = b + c ↔ a + c = b := by
+  rw [← eq_sub_iff_add_eq, sub_eq_add]
+
+@[scoped simp]
+protected theorem two_zsmul (x : R) : (2 : ℤ) • x = 0 := by
+  rw [two_zsmul, add_self_eq_zero]
```
Comment (Command-Master) [inline Mathlib/Algebra/CharP/Two.lean:79]: Perhaps also add a `two_smul` version for an arbitrary module, like [two_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Defs.html#two_smul)?

File: Mathlib/Algebra/CharP/Two.lean [hunk 3]
```diff
@@ -20,15 +20,34 @@ variable {R ι : Type*}
...
+  rw [← Nat.cast_two, CharP.cast_eq_zero]
+
+/-- The only hypotheses required to build a `CharP R 2` instance are `1 ≠ 0` and `2 = 0`. -/
+protected theorem mk (h₁ : (1 : R) ≠ 0) (h₂ : (2 : R) = 0) : CharP R 2 := by
+  refine ⟨fun n ↦ ?_⟩
+  obtain hn | hn := Nat.even_or_odd n
+  · simp_rw [hn.two_dvd, iff_true]
+    exact natCast_eq_zero_of_even_of_two_eq_zero hn h₂
+  · simp_rw [hn.not_two_dvd_nat, iff_false]
+    rwa [natCast_eq_one_of_odd_of_two_eq_zero hn h₂]
```
Comment (adomani) [inline Mathlib/Algebra/CharP/Two.lean:37]:
```suggestion
protected theorem mk (h₁ : (1 : R) ≠ 0) (h₂ : (2 : R) = 0) : CharP R 2 where
  cast_eq_zero_iff' n := by
    obtain hn | hn := Nat.even_or_odd n
    · simp_rw [hn.two_dvd, iff_true]
      exact natCast_eq_zero_of_even_of_two_eq_zero hn h₂
    · simp_rw [hn.not_two_dvd_nat, iff_false]
      rwa [natCast_eq_one_of_odd_of_two_eq_zero hn h₂]
```
This syntax is also available: I prefer it, but I do not think that there is a clear style guidance either way.

Comment (vihdzp) [reply Mathlib/Algebra/CharP/Two.lean:37]: Sure, I don't mind writing it like this.

File: Mathlib/Algebra/CharP/Two.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/Algebra/CharP/Two.lean]: This is meant as a pseudo-constructor of sorts. I think the name is fine, since I don't think there's going to be any alternate ways of building the instance that don't just boil down to this, and alternate names like `char_two_of_one_ne_zero_of_two_eq_zero` are way too wordy.

File: Mathlib/Algebra/CharP/Two.lean [hunk 4]
```diff
@@ -37,16 +56,27 @@ variable [Ring R] [CharP R 2]
...
   funext neg_eq
 
-@[scoped simp]
 theorem sub_eq_add (x y : R) : x - y = x + y := by rw [sub_eq_add_neg, neg_eq]
 
-theorem sub_eq_add' : HSub.hSub = ((· + ·) : R → R → R) :=
-  funext fun x => funext fun y => sub_eq_add x y
+/-- `-` and `+` are the same function in characteristic 2. -/
+@[scoped simp]
+theorem sub_eq_add' : HSub.hSub = (HAdd.hAdd : R → R → R) :=
```
[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/Algebra/CharP/Two.lean:68]: I wanted to make sure that every single instance of `-` got turned into `+` , though I guess there aren't actually very many instances where `-` is unapplied. Should I just deprecate this lemma?

File: Mathlib/Algebra/CharP/Two.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/Algebra/CharP/Two.lean]: Shouldn't it be `of_one_ne_zero_of_two_eq_zero`? I guess this is still less wordy than I was expecting, so I can perform the rename.

File: Mathlib/Algebra/CharP/Two.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/CharP/Two.lean:68]: Yes, it appears unused.

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/CharP/Two.lean:68]: Except in `scripts/no_lints_prime_decls.txt`

File: Mathlib/Algebra/CharP/Two.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/CharP/Two.lean]: Do we need two `of`'s?

File: Mathlib/Algebra/CharP/Two.lean (hunk 2 above)
Comment (vihdzp) [reply Mathlib/Algebra/CharP/Two.lean:79]: Does that work? `two_smul` takes an argument `x : M` in the monoid, rather than `x : R` in the characteristic two semiring.

Comment (mattrobball) [reply Mathlib/Algebra/CharP/Two.lean:81]: I think that is a reasonable request but let's save it for another PR. @Command-Master feel free to PR such a change and ping me.

**Comment by mattrobball** [comment]
bors merge
